react/hooks/prefetch.ts

```diff
@@ -0,0 +1,266 @@
+        declarer,
+        hints,
+        renderMajor,
+      }),
+    }
+    newState.promise
+      .then((data) => {
+        routesCache.set(navigationPage, data)
+      })
+      .finally(() => {
+        newState.promisePending = false
+        newState.promise = null as any
+        routePromise[navigationPage] = null
+      })
+
+    routePromise[navigationPage] = newState
+  }
+  const globalRouteDataState = routePromise[navigationPage]
+  if (globalRouteDataState?.promisePending) {
+    await globalRouteDataState.promise
+    routeDataCache = routesCache.get(navigationPage)
+  }
+
+  if (!routeDataCache) {
+    return
+  }
+
+  await fetchComponents(
+    routeDataCache.components,
+    window.__RUNTIME__,
+    routeDataCache.extensions
+  )
+}
+
+interface UsePrefetchArgs {
+  inView: boolean
+  page?: string
+  href: string
+  options: NavigateOptions
+  waitToPrefetch?: number
+}
+
+const getCacheValidData = (path: string, prefecthState: PrefetchState) => {
+  const { pathsState, routesCache } = prefecthState
+  const page = pathsState[path]?.page
+
+  if (!page) {
+    return {
+      pathValid: false,
+      routeValid: false,
+    }
+  }
+
+  const cache = getCacheForPage(page)
+  const validData = cache.has(path)
+  const validRoutesData = routesCache.has(page)
+  return {
+    pathValid: validData,
+    routeValid: validRoutesData,
+  }
+}
+
+export const usePrefetchAttempt = ({
+  inView,
+  page,
+  href,
+  options,
+  waitToPrefetch,
+}: UsePrefetchArgs) => {
+  const runtime = useRuntime()
+  const prefetchState = usePrefetch()
+  const client = useApolloClient() as ApolloClientType
+  const hasTried = useRef(false)
+  const [canPrefetch, setCanPrefetch] = useState(() => {
+    return !waitToPrefetch || waitToPrefetch === 0
+  })
+
+  useEffect(() => {
+    if (!canPrefetch) {
+      setTimeout(() => {
+        setCanPrefetch(true)
+      }, waitToPrefetch)
+    }
+  }, [canPrefetch, waitToPrefetch])
+
+  const { pages, navigationRouteModifiers, hints, renderMajor } = runtime
+
+  useEffect(() => {
+    if (inView && !hasTried.current && canPrefetch) {
+      const { pathsState, queue } = prefetchState
+      if (href && href[0] !== '/') {
+        // Should only work on relative paths
+        return
+      }
+
+      options.modifiers = navigationRouteModifiers
+      const navigationRoute = getNavigationRouteToNavigate(pages, options)
+      navigationRoute.original = options.to
+
+      if (pathsState[navigationRoute.path]?.fetching) {
+        // already fetching, no need to go any further
+        return
+      }
+
+      const validCache = getCacheValidData(navigationRoute.path, prefetchState)
+
+      if (validCache.pathValid && validCache.routeValid) {
+        // cache all valid, no need to fetch anything
+        return
+      }
+
+      if (!validCache.pathValid) {
+        pathsState[navigationRoute.path] = { fetching: true }
+      }
+
+      const priority = page === 'store.product' ? 1 : 0
```

Comment:
this is kind of weird to be hardcoded here inside this function. could this be a separate function? (as in, `getPagePriority(page)`)